src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.cs

```diff
@@ -556,6 +540,216 @@ public void Add(NamespaceOrTypeSymbol symbol)
+                        }
+                        return;
+                }
+
+                InjectedNamespaceSymbol makeEmptyNamespace(string name)
+                {
+                    return new InjectedNamespaceSymbol(containingNamespace.ContainingModule, containingNamespace, name);
+                }
+
+                bool isSystem(NamespaceSymbol symbol)
+                {
+                    return symbol.Name == system && symbol.ContainingNamespace.IsGlobalNamespace;
+                }
+
+                bool isRuntime(NamespaceSymbol symbol)
+                {
+                    return symbol.Name == runtime && isSystem(symbol.ContainingNamespace);
+                }
+
+                bool isMicrosoft(NamespaceSymbol symbol)
+                {
+                    return symbol.Name == microsoft && symbol.ContainingNamespace.IsGlobalNamespace;
+                }
+            }
+
+            class InjectedNamespaceSymbol : NamespaceSymbol
+            {
+                private ModuleSymbol _module;
+                private NamespaceSymbol _containingNamespace;
+                private string _name;
+                private Dictionary<string, ImmutableArray<NamespaceOrTypeSymbol>> _nameToMembersMap;
+                private Dictionary<string, ImmutableArray<NamedTypeSymbol>> _nameToTypeMembersMap;
+                private ImmutableArray<Symbol> _lazyAllMembersOrdered;
+                private ImmutableArray<NamedTypeSymbol> _lazyTypeMembersOrdered;
+
+                public InjectedNamespaceSymbol(ModuleSymbol module, NamespaceSymbol containingNamespace, string name)
+                {
+                    Debug.Assert(name != null);
+                    _module = module;
+                    _containingNamespace = containingNamespace;
+                    _name = name;
+                }
+
+                public override string Name 
+                    => _name;
+
+                public override AssemblySymbol ContainingAssembly
+                    => _containingNamespace.ContainingAssembly;
+
+                public override Symbol ContainingSymbol
+                    => _containingNamespace;
+
+                public override ImmutableArray<Location> Locations
+                    => ImmutableArray<Location>.Empty;
+
+                public override ImmutableArray<SyntaxReference> DeclaringSyntaxReferences
+                    => ImmutableArray<SyntaxReference>.Empty;
+
+                internal override NamespaceExtent Extent
+                    => new NamespaceExtent(_module);
+
+                internal override ImmutableArray<Symbol> GetMembersUnordered()
+                {
+                    var result = GetMembers();
+
+#if DEBUG
+                    // In DEBUG, swap first and last elements so that use of Unordered in a place it isn't warranted is caught
+                    // more obviously.
+                    return result.DeOrder();
+#else
+                    return result;
+#endif
```

Comment:
This pattern is used many places. Should we add a `ConditionallyDeOrder()` method that does this?